### ISSUE TITLE
How do I force Jellyfin to use HEVC for transcoding instead of H264?

### ISSUE THREAD

**Issue by avc1657** [jellyfin/jellyfin-tizen#134]

**Comment by dmitrylyzo** [comment]
I don't think we have the implemented feature of preferring one codec over another. This can be done by changing the order of codecs in the container in the device profile before use.
https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L531
https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L535

Maybe some kind of postprocessing of the device profile as here https://github.com/jellyfin/jellyfin-web/blob/7771ce716e4eab22f17a3354b9b0bd8e94e9291a/src/components/apphost.js#L42-L60

**Comment by avc1657** [comment]
You mean doing changes here? If so, what changes?

![msedge_tQ5L4gShSm](https://user-images.githubusercontent.com/68762165/187027554-603e581a-4bf0-4553-b526-3ff5914f75e7.png)

**Comment by dmitrylyzo** [comment]
No, these are DLNA settings (without the app). I mean, you need to modify the code of jf-web because no one has implemented this feature yet.

**Comment by avc1657** [comment]
Oh, and what changes do I do? Whenever transcoding is needed Jellyfin uses H264 as the output codec but my TV supports H265. I want H265 as the codec.

**Comment by dmitrylyzo** [comment]
Quick way:
Move block https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L528-L537 before https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L519

**Comment by avc1657** [comment]
Thank you very much

**Comment by avc1657** [comment]
How do i make so jellyfin transcodes using 10 bits instead of 8? Will writing "hevc main10" work?

**Comment by dmitrylyzo** [comment]
`main10` is codec profile and it should be already specified for Tizen.
_Full device profile is logged on the server. If it is not, change Jellyfin logging to `Debug`._

I don't think you can force 10-bit. Server should probably look at the color space of the source media and activate it if necessary.

**Comment by avc1657** [comment]
> `main10` is codec profile and it should be already specified for Tizen. _Full device profile is logged on the server. If it is not, change Jellyfin logging to `Debug`._
> 
> I don't think you can force 10-bit. Server should probably look at the color space of the source media and activate it if necessary.

Most of the time the original media info says something like HEVC Main 10, but the transcoded info only says HEVC.
This JS file has some parts related to main10: browserDeviceProfile.js 
Maybe i should edit there? Idk.

**Comment by avc1657** [comment]
Even if i disable HEVC, Jellyfin uses 8bit for H264 even tho the source is HEVC Main 10. Yesterday I watched a show and sometimes the color banding because of 8bit looked sooo bad. Well, maybe the cause is not exactly the 8bit transcoding, idk. The source video does not have color banding in the same scenes.

https://github.com/jellyfin/jellyfin/issues/5576 -> sorry i dont understand how this can be my issue. 

Btw it's not like i need HEVC, i just don't see why would i use H264 over HEVC if my GPU supports HEVC encoding. H264 is older, less efficient, i don't see why i would use it.

https://github.com/jellyfin/jellyfin/issues/5576#issuecomment-806178759 -> then i should edit something here?

https://github.com/jellyfin/jellyfin/pull/9121 -> idk if i get this correctly. Let's say the original video file is 2000kbps hevc main 10. So the transcoded media would still use 2000kbps for encoding? If so, that's quite bad. Would lose too much quality. The idea i follow is to set CRF to 20 so it uses 2x/3x more bitrate so transcoded media doesnt lose much quality. 

Im definitely not sure of what should i do.

**Comment by avc1657** [comment]
> Btw, is the original video SDR or HDR? The issue only happens for a HEVC HDR source when it is supported by the client (copy stream).

The original video is SDR

> AFAIK, even if you have enabled HEVC encoding on the server, the latter will use the first in the list of supported codecs ([#134 (comment)](https://github.com/jellyfin/jellyfin-tizen/issues/134#issuecomment-1229191012)).
> Client-side reordering (via UI) would be useful, but it is not implemented.

Seems to be the case, yeah. I changed the block ordering because even after i enabled HEVC, the transcoding was using H264. Cant i edit the hevc block (https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L528-L537) telling to use hevc main 10?

**Comment by dmitrylyzo** [comment]
> The original video is SDR

Then it is not caused by that block.

What is the reason for transcoding (`Playback Info` on the TV or on the Dashboard)?
_Burning subs probably (https://github.com/jellyfin/jellyfin-tizen/issues/130#issuecomment-1208136820)._

> telling to use hevc main 10?

Supported profiles are listed in https://github.com/jellyfin/jellyfin-web/blob/00fdadd1ced05b542cd07bd253bc234a1e214369/src/scripts/browserDeviceProfile.js#L843 and below. But I don't think reordering will help.

Iirc, you can find the requested stream (URL) in the Ffmpeg log. Something like `main.m3u8?...hevc-profile=...`.

**Comment by avc1657** [comment]
> Burning subs probably ([#130 (comment)](https://github.com/jellyfin/jellyfin-tizen/issues/130#issuecomment-1208136820)).

Yeah, burning subs.

First of all: i checked the transcode folder from which transcoded media comes when im playing stuff on my TV. Right clicked any TS  video file and clicked to show info with media info. Media info clearly said HEVC Main L4, not Main 10. So you said Jellyfin is supposed to check the original source and transcode accordingly, which doesn't happen. 90%+ of my media are 10bit, including the one in this example, yet jellyfin is transcoding using 8-bit.

Ok, so:

int main () {

    If (there's a solution for FORCE my TV to ALWAYS use HEVC Main 10, no matter if original video is 8bit) {
        could you please tell me what lines of code to edit and what should i type in then in order to produce the wanted result?
    } else {
         It can't be helped, then. If there's no way, there's no way.
    }
return 0;
}

Thks for you time helping me :) .
